keyvault/shim.go

```diff
@@ -0,0 +1,158 @@
+package keyvault
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/keyvault/azsecrets"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/pkcs12"
+)
+
+const (
+	pemContentType    = "application/x-pem-file"
+	pkcs12ContentType = "application/x-pkcs12"
+)
+
+type secretFetcher interface {
+	GetSecret(ctx context.Context, secretName string, opts *azsecrets.GetSecretOptions) (azsecrets.GetSecretResponse, error)
+}
+
+// Shim provides convenience methods for working with KeyVault.
+type Shim struct {
+	sf secretFetcher
+}
+
+// NewShim constructs a Shim for a KeyVault instance located at the provided url. The azcore.TokenCredential will
+// only be used during method calls, it is not verified at initialization.
+func NewShim(vaultURL string, cred azcore.TokenCredential) (*Shim, error) {
+	c, err := azsecrets.NewClient(vaultURL, cred, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create new azsecrets.Client")
+	}
+
+	return &Shim{sf: c}, nil
+}
+
+// GetLatestTLSCertificate fetches the latest version of a keyvault certificate and transforms it into a usable tls.Certificate.
+func (s *Shim) GetLatestTLSCertificate(ctx context.Context, certName string) (tls.Certificate, error) {
+	resp, err := s.sf.GetSecret(ctx, certName, nil)
+	if err != nil {
+		return tls.Certificate{}, errors.Wrap(err, "could not get secret")
+	}
+
+	pemBlocks, err := getPEMBlocks(*resp.Properties.ContentType, *resp.Value)
```

Comment:
Does PEM follow MIME to the letter? Like, could you end up with surprise things after the media type (e.g., as absurd as this would be: `application/x-pem-file; charset=utf-8`. There's a `mime.ParseMediaType` that does the right thing, if so.

Reply:
For keyvault-generated certs (our target use case), what I've observed are these specific content types, though [documentation about kv secrets is not totally clear](https://docs.microsoft.com/en-us/azure/key-vault/secrets/about-secrets).